nav2_mppi_controller/include/nav2_mppi_controller/tools/utils.hpp

```diff
@@ -612,6 +606,80 @@ struct Pose2D
   float x, y, theta;
 };
 
+/**
+ * @brief Shift the columns of a 2D Eigen Array or scalar values of
+ *    1D Eigen Array by 1 place.
+ * @param e Eigen Array
+ * @param direction direction in which Array will be shifted.
+ *     1 for shift in right direction and -1 for left direction.
+ */
+inline void shiftColumnsByOnePlace(Eigen::Ref<Eigen::ArrayXXf> e, int direction)
+{
+  int size = e.size();
+  if(size == 1) {return;}
+  if((e.cols() == 1 || e.rows() == 1) && size > 1) {
+    auto start_ptr = direction == 1 ? e.data() + size - 2 : e.data() + 1;
+    auto end_ptr = direction == 1 ? e.data() : e.data() + size - 1;
+    while(start_ptr != end_ptr) {
+      *(start_ptr + direction) = *start_ptr;
+      start_ptr -= direction;
+    }
+    *(start_ptr + direction) = *start_ptr;
+  } else {
+    auto start_ptr = direction == 1 ? e.data() + size - 2 * e.rows() : e.data() + e.rows();
+    auto end_ptr = direction == 1 ? e.data() : e.data() + size - e.rows();
+    auto span = e.rows();
+    while(start_ptr != end_ptr) {
+      std::copy(start_ptr, start_ptr + span, start_ptr + direction * span);
+      start_ptr -= (direction * span);
+    }
+    std::copy(start_ptr, start_ptr + span, start_ptr + direction * span);
+  }
+}
+
+/**
+ * @brief Normalize the yaws between points on the basis of final yaw angle
+ *    of the trajectory.
+ * @param last_yaws Final yaw angles of the trajectories.
+ * @param yaw_between_points Yaw angles calculated between x and y co-ordinates of the trajectories.
+ * @return Normalized yaw between points.
+ */
+inline auto normalize_yaws_between_points(
+  const Eigen::Ref<const Eigen::ArrayXf> & last_yaws,
+  const Eigen::Ref<const Eigen::ArrayXf> & yaw_between_points)
+{
+  Eigen::ArrayXf yaws = utils::shortest_angular_distance(
+          last_yaws, yaw_between_points).abs();
+  int size = yaws.size();
+  Eigen::ArrayXf yaws_between_points_corrected(size);
+  for(int i = 0; i != size; i++) {
+    const float & yaw_between_point = yaw_between_points[i];
+    yaws_between_points_corrected[i] = yaws[i] < M_PIF_2 ?
+      yaw_between_point : angles::normalize_angle(yaw_between_point + M_PIF);
+  }
+  return yaws_between_points_corrected;
```

Comment:
Can we remove this copy?

Reply:
Sorry, I didn't get it. Are you referring about copy during return?

Reply:
Yes

Reply:
Will it be copied though? I think it won't be copied due to the named return-value optimization by compiler, since Eigen::Array has a defined Move constructor.